### DISCUSSION TITLE
'Private' functions

### DISCUSSION THREAD

**Discussion by dolgorukee** [supabase/supabase#13107] [Questions] [answered]
Hi,

Is there a way to make a function 'private' in the sense that it can only be executed from the server and not remotely via JS (for example).

I used `ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;` and I can see my function has `REVOKE ALL ON FUNCTION public.test() FROM PUBLIC;` but there are still executes granted for other roles:
```
GRANT EXECUTE ON FUNCTION public.test() TO anon;
GRANT EXECUTE ON FUNCTION public.test() TO authenticated;
GRANT EXECUTE ON FUNCTION public.test() TO postgres;
GRANT EXECUTE ON FUNCTION public.test() TO service_role;
```

**Comment by GaryAustin1** [comment, marked as answer]
You can revoke anon (for anon users), authenticated for signed in users and service_role if you don't want that key to work either.
Or could just create another schema, not enable it for the API, and put your "special" functions there.
